cmd/cellrangerBam/cellrangerBam.go

```diff
@@ -0,0 +1,823 @@
+	ncNorm          string
+	determineBins   string
+	inputSequencing string
+}
+
+// calculateNormFactor takes the countsMap created in readInputSequencingSam and writes out a data frame with the name of the construct, counts per 500bp, percent abundance, and normalization factor
+func calculateNormFactor(countsMap map[string]float64, outFile string) {
+	var totalReads float64 = 0
+	var line string
+	var percentLib float64
+	var matrix []string
+
+	out := fileio.EasyCreate(outFile)
+	fileio.WriteToFileHandle(out, "construct\treadsPer500bp\tpercentLibrary\tnormFactor")
+
+	numConstructs := len(countsMap)
+	idealPerc := 100.0 / float64(numConstructs)
+
+	for i := range countsMap {
+		totalReads = countsMap[i] + totalReads
+	}
+
+	for i := range countsMap {
+		percentLib = (countsMap[i] / totalReads) * 100
+		line = fmt.Sprintf("%s\t%f\t%f\t%f", i, countsMap[i], percentLib, idealPerc/percentLib)
+		matrix = append(matrix, line)
+	}
+	sort.Strings(matrix)
+	for _, i := range matrix {
+		fileio.WriteToFileHandle(out, i)
+	}
+	err := out.Close()
+	exception.PanicOnErr(err)
+}
+
+// readInputSequencingSam reads and parses an alignment file for an input library sequencing run. It creates a read-count map for all constructs in the provided GTF
+func readInputSequencingSam(s Settings) {
+	var tree = make([]interval.Interval, 0)
+	var bedSizeMap = make(map[string]int)
+	var countsMap = make(map[string]float64)
+	var selectTree map[string]*interval.IntervalNode
+	var constructName string
+	var counts float64
+	var currEntry sam.Sam
+
+	inChan, _ := sam.GoReadToChan(s.inFile)
+	bedEntries := bed.Read(s.inputSequencing)
+
+	for _, i := range bedEntries {
+		tree = append(tree, i)
+		bedSizeMap[i.Name] = bed.Size(i)
+		countsMap[i.Name] = 0
+	}
+	selectTree = interval.BuildTree(tree)
+
+	currEntry = <-inChan
+	for i := range inChan {
+		//remove PCR duplicates
+		if interval.AreEqual(currEntry, i) {
+			if currEntry.Qual >= i.Qual {
+				continue
+			} else {
+				currEntry = i
+				continue
+			}
+		}
+		overlappedBedEntry := interval.Query(selectTree, currEntry, "any")
+		if len(overlappedBedEntry) != 1 {
+			currEntry = i
+			continue
+		}
+		constructName = overlappedBedEntry[0].(bed.Bed).Name
+		counts, _ = countsMap[constructName]
+		countsMap[constructName] = counts + 1
+		currEntry = i
+	}
+	//normalize reads to 500bp
+	for i := range countsMap {
+		counts, _ = countsMap[i]
+		countsMap[i] = counts / (float64(bedSizeMap[i]) / 500.0)
+	}
+	calculateNormFactor(countsMap, s.outFile)
+}
+
+// getNcAvg takes in the input normalized matrix of single cell data, the list of negative control sequences and an empty slice of slices and will return the average negative control reads for each
+// cell type
+func getNcAvg(matrix []string, ncSlice []string, ncVals [][]float64) []float64 {
+	var found bool
+	var currCluster string
+	var columns []string
+	var prevCluster string
+	var ncAvg []float64
+
+	columns = strings.Split(matrix[0], "\t")
+	currCluster = columns[0]
```

Comment:
Yeah good question i don't exactly know. My reasoning was: I wanted to set the first entry to "currCluster" because i have logic in the for loop that compares currCluster to prevCluster. If they are both nil then on the first time through the loop the will be equal (?) and move onto the next cluster when i don't want it to